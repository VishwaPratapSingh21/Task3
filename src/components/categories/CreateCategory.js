import React, { Component } from "react";
import { notify } from "react-notify-toast";
import axiosInstance from "../Apicalls";

/**
 * Component for allowing users to add categories.
 * @author [Hadijah kyampeire](https://github.com/hadijahkyampeire/Yummy_Reactjs_frontend)
 */
class CreateCategory extends Component {
  state = {
    title: ""
  };

  handleInputChange = event => {
    const { name, value } = event.target;
    this.setState({ [name]: value });
  };
  /**
   * handles addition of categories
   * @param {object} event - Reference to the submit form component
   * used to prevent default form behaviours on submission
   */
  handleAddCategories = (event, id) => {
    const { title: name } = this.state;
    event.preventDefault();
    axiosInstance
      .post("categories/", { name })
      .then(response => {
        notify.show(response.data.message, 'success', 4000);
        document.getElementById("closeAddModal").click();

        this.props.getCategories();
        this.setState({ title: "" });
      })
      .catch(error => {
        if (error.response) {
          this.setState({ error: error.response.data.message, mess: "" });
        } else if (error.request) {
          notify.show("Request not made");
        }
      });
  };

  render() {
    return (
      <div>
        <button
          type="button"
          className="btn btn-primary"
          data-toggle="modal"
          data-target="#createCategoryModal"
          aria-hidden="true"
        >
          <i className="fa fa-plus" /> Add category
        </button>
        <div className="modal fade" id="createCategoryModal">
          <div className="modal-dialog" role="document">
            <div className="modal-content">
              <div className="modal-header">
                <h5 className="modal-title">Add category</h5>
                <button
                  type="button"
                  className="close"
                  data-dismiss="modal"
                  aria-label="Close"
                >
                  <span aria-hidden="true">&times;</span>
                </button>
              </div>
              <form onSubmit={this.handleAddCategories}>
                <div className="modal-body">
                  {this.state.error ? (
                    <div className="alert alert-danger">{this.state.error}
                     <button
                    type="button"
                    class="close"
                    data-dismiss="alert"
                    aria-label="Close"
                  >
                    <span aria-hidden="true">&times;</span>
                  </button></div>
                  ) : (
                    ""
                  )}

                  <div className="input-group ">
                    <span className="input-group-addon pr-4">
                      <i className="glyphicon glyphicon-pencil" />
                    </span>
                    <input
                      type="text"
                      name="title"
                      className="form-control"
                      placeholder="category name"
                      onChange={this.handleInputChange}
                      value={this.state.title}
                    />
                  </div>
                </div>
                <div className="modal-footer">
                  <input
                    type="submit"
                    className="btn btn-primary"
                    value="Add"
                  />
                  <button
                    type="button"
                    className="btn btn-secondary"
                    data-dismiss="modal"
                    id="closeAddModal"
                  >
                    Close
                  </button>
                </div>
              </form>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
export default CreateCategory;
