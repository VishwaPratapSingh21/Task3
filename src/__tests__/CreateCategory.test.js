import React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import toJson, { shallowToJson } from 'enzyme-to-json';
import sinon from 'sinon';

import CreateCategory from '../components/categories/CreateCategory';

describe('CreateCategory component', () => {
  const wrapper = shallow(<CreateCategory />);
  const preventDefault = jest.fn();

  it('renders properly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });

  it('has initial state', () => {
    expect(wrapper.state().title).toEqual('');
  });

  it('changes state', () => {
    wrapper.setState({ title: 'lunch' });
    expect(wrapper.find('[name="title"]').props().value).toEqual('lunch');
  });

  it('has the correct form fields', () => {
    expect(wrapper.find('[name="title"]')).toHaveLength(1);
  });

  it('renders component in divs', () => {
    expect(wrapper.find('div')).toHaveLength(8);
  });

  it('renders component in spans', () => {
    expect(wrapper.find('span')).toHaveLength(2);
  });

  it('has input field', () => {
    expect(wrapper.find('input')).toHaveLength(2);
  });

  it('renders a form', () => {
    expect(wrapper.find('form')).toHaveLength(1);
    expect(wrapper.find('form').simulate('submit', { preventDefault }));
    expect(preventDefault).toBeCalled();
    // expect(wrapper.find('[name="name"]').simulate('change', {target:{name:'name', value:'lunch'}}))
  });
});
