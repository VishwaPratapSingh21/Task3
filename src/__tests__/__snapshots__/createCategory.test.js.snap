// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCategory component renders properly 1`] = `
<div>
  <button
    aria-hidden="true"
    className="btn btn-primary"
    data-target="#createCategoryModal"
    data-toggle="modal"
    type="button"
  >
    <i
      className="fa fa-plus"
    />
     Add category
  </button>
  <div
    className="modal fade"
    id="createCategoryModal"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header"
        >
          <h5
            className="modal-title"
          >
            Add category
          </h5>
          <button
            aria-label="Close"
            className="close"
            data-dismiss="modal"
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <form
          onSubmit={[Function]}
        >
          <div
            className="modal-body"
          >
            <div
              className="input-group "
            >
              <span
                className="input-group-addon pr-4"
              >
                <i
                  className="glyphicon glyphicon-pencil"
                />
              </span>
              <input
                className="form-control"
                name="title"
                onChange={[Function]}
                placeholder="category name"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            className="modal-footer"
          >
            <input
              className="btn btn-primary"
              type="submit"
              value="Add"
            />
            <button
              className="btn btn-secondary"
              data-dismiss="modal"
              id="closeAddModal"
              type="button"
            >
              Close
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
